### PR TITLE
Adapt to OpenAPI3

### LINKED ISSUE #28: Implement OAPI spec generator/server for aiohttp
Best implemented after #23

### LINKED ISSUE #29: Refactor Flask and AioHTTP extensions
AioHTTP extension should not extend the Flask extension, instead there should be a common ancestor for these.

### PR DESCRIPTION
**Pull request by Teyras**
closes #28 
closes #29

### CODE REVIEW

File: apistrap/aiohttp.py [hunk 1]
```diff
@@ -198,18 +204,107 @@ def _handle_http_error(self, exception):
...
+        if isinstance(route.resource, DynamicResource):
+            parameters = re.findall(r"{([a-zA-Z0-9]+[^}]*)}", route.resource.get_info()["formatter"])
+
+            for arg in parameters:
+                param_data = {
+                    "in": "path",
+                    "name": arg,
+                    "required": True,
+                    "schema": {
+                        "type": "string"  # TODO support other types too
```
Comment (blazekadam) [inline apistrap/aiohttp.py]: ha...

File: apistrap/flask.py [hunk 2]
```diff
@@ -1,43 +1,178 @@
...
+        :param app: the Flask instance
         """
+        self._app = app
+        blueprint = Blueprint("apistrap", __name__, template_folder=path.join(path.dirname(__file__), "templates"))
+
+        if self.spec_url is not None:
+            blueprint.route(self.spec_url, methods=["GET"])(self._get_spec)
+
+        if self.spec_url is not None and self.ui_url is not None:
+            blueprint.route(self.ui_url)(self._get_ui)
```
Comment (blazekadam) [inline apistrap/flask.py]: ditto ending with `/`

File: apistrap/flask.py [hunk 3]
```diff
@@ -1,43 +1,178 @@
...
+                if method.lower() not in ["get", "post", "put", "delete", "patch"]:
+                    continue
+
+                self.spec.path(url, {
+                    method.lower(): self._extract_operation_specs(handler)
+                })
+
+        self._specs_extracted = True
+
+    def _extract_operation_specs(self, handler):
```
Comment (blazekadam) [inline apistrap/flask.py]: This is awkwardly similar to the the implementation in `AioHTTPApistrap`...

File: apistrap/flask.py [hunk 4]
```diff
@@ -1,43 +1,178 @@
...
+        """
+        Extract operation specification data from a Flask view handler
+        :param handler: the Flask handler to extract
```
Comment (blazekadam) [inline apistrap/flask.py]: Missing newline

File: apistrap/flask.py [hunk 5]
```diff
@@ -1,43 +1,178 @@
...
+        :return: a dictionary containing the specification data
+        """
+
```
Comment (blazekadam) [inline apistrap/flask.py]: Extra newline :)

File: apistrap/aiohttp.py (hunk 1 above)
Comment (Teyras) [reply apistrap/aiohttp.py]: this will have to wait until another PR

File: apistrap/flask.py (hunk 3 above)
Comment (Teyras) [reply apistrap/flask.py]: It is alright. I vote we wait for this to flesh out a bit more and then refactor it.

File: apistrap/flask.py (hunk 2 above)
Comment (Teyras) [reply apistrap/flask.py]: ? we do that on the line below

File: apistrap/aiohttp.py [hunk 6]
```diff
@@ -198,18 +204,107 @@ def _handle_http_error(self, exception):
...
         self.app = app
         app.middlewares.append(self.error_middleware)
 
-    def responds_with(self, response_class: Type[Model], *, code: int=200, description: Optional[str]=None,
-                      mimetype: Optional[str]=None):
+        if self.spec_url is not None:
+            app.router.add_route("get", self.spec_url, self._get_spec)
+
+        if self.ui_url is not None:
+            for ui_url in (self.ui_url, self.ui_url + "/"):
```
[the remark this answers is not in the document]
Comment (Teyras) [reply apistrap/aiohttp.py]: It doesn't, there is a .rstrip() call in the setter.

File: apistrap/flask.py (hunk 3 above)
Comment (blazekadam) [reply apistrap/flask.py]: Roger that

File: apistrap/aiohttp.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (blazekadam) [reply apistrap/aiohttp.py]: Ah, sorry I missed that...

File: apistrap/errors.py [hunk 7]
```diff
@@ -1,9 +1,9 @@
-from typing import List, Dict
+from typing import Dict, List
 
 
-class SwaggerExtensionError(Exception):
+class ApistrapExtensionError(Exception):
```
Comment (blazekadam) [inline apistrap/errors.py]: The word *Extension* could have been omitted.

**Pull request merged**